community/kernel/src/main/java/org/neo4j/kernel/recovery/LatestCheckPointFinder.java

```diff
@@ -91,9 +93,28 @@ public LatestCheckPoint find( long fromVersionBackwards ) throws IOException
 
             if ( latestCheckPoint != null )
             {
+                // Is the latest start entry in this log file version later than what the latest check point targets?
+                LogPosition target = latestCheckPoint.getLogPosition();
                 boolean commitsAfterCheckPoint = latestStartEntry != null &&
-                        latestStartEntry.getStartPosition().compareTo( latestCheckPoint.getLogPosition() ) >= 0;
-                return new LatestCheckPoint( latestCheckPoint, commitsAfterCheckPoint , oldestVersionFound );
+                        latestStartEntry.getStartPosition().compareTo( target ) >= 0;
+                if ( !commitsAfterCheckPoint )
+                {
+                    if ( target.getLogVersion() < version )
+                    {
+                        // This check point entry targets a previous log file.
+                        // Go there and see if there's a transaction. Reader is capped to that log version.
+                        commitsAfterCheckPoint = extractFirstTxIdAfterPosition( target, target.getLogVersion() ) !=
+                                LatestCheckPoint.NO_TRANSACTION_ID;
+                    }
+                }
+
+                // Extract first transaction id after check point target position.
+                // Reader may continue into log files after the initial version.
+                long firstTxIdAfterCheckPoint = commitsAfterCheckPoint
+                        ? extractFirstTxIdAfterPosition( target, fromVersionBackwards )
+                        : LatestCheckPoint.NO_TRANSACTION_ID;
+                return new LatestCheckPoint( latestCheckPoint, commitsAfterCheckPoint,
+                        firstTxIdAfterCheckPoint, oldestVersionFound );
```

Comment:
Can we simplify the logic in this whole block to something like:

```
if ( latestCheckPoint != null )
{
    // Is the latest start entry in this log file version later than what the latest check point targets?
    LogPosition target = latestCheckPoint.getLogPosition();

    long firstTxAfterCheckPoint = LatestCheckPoint.NO_TRANSACTION_ID;
    if ( latestStartEntry != null )
    {
        // We found a start entry in some of the log versions from current (including) or later
        firstTxAfterCheckPoint = extractFirstTxIdAfterPosition( target, fromVersionBackwards );
    }
    else if ( target.getLogVersion() < version )
    {
        // We did not find any start entries in current version (including) or later
        // We can safely cap reader to only read up to version previous to current
        firstTxAfterCheckPoint = extractFirstTxIdAfterPosition( target, version - 1 );
    }
    boolean commitsAfterCheckPoint = firstTxAfterCheckPoint != LatestCheckPoint.NO_TRANSACTION_ID;
    return new LatestCheckPoint( latestCheckPoint, commitsAfterCheckPoint, firstTxAfterCheckPoint, oldestVersionFound );
}
```

I tried it but it breaks `LatestCheckPointFinderTest.latestLogFileContainingACheckPointAndAStartWithoutCommitAfter` which I suspect is wrong, see comment below.

In any case, it would be super nice to simplify this block.